sklearn/metrics/pairwise.py

```diff
@@ -231,34 +231,115 @@ def euclidean_distances(X, Y=None, Y_norm_squared=None, squared=False,
         elif XX.shape != (X.shape[0], 1):
             raise ValueError(
                 "Incompatible dimensions for X and X_norm_squared")
+        if XX.dtype == np.float32:
+            XX = None
+    elif X.dtype == np.float32:
+        XX = None
     else:
         XX = row_norms(X, squared=True)[:, np.newaxis]
 
-    if X is Y:  # shortcut in the common case euclidean_distances(X, X)
+    if X is Y and XX is not None:
+        # shortcut in the common case euclidean_distances(X, X)
         YY = XX.T
     elif Y_norm_squared is not None:
         YY = np.atleast_2d(Y_norm_squared)
 
         if YY.shape != (1, Y.shape[0]):
             raise ValueError(
                 "Incompatible dimensions for Y and Y_norm_squared")
+        if YY.dtype == np.float32:
+            YY = None
+    elif Y.dtype == np.float32:
+        YY = None
```

Comment:
At first glance, it is not obvious that `None` is used by `_euclidean_distances_upcast`.

Reply:
What do you mean ? `XX` and `YY` are used by `_euclidean_distances_upcast` no matter their value

Reply:
Setting `XX` or `YY` to `None` tells `_euclidean_distances_upcast` to call `row_norms(*_chunk, squared=True)[:, np.newaxis]`. 

Without reading `_euclidean_distances_upcast`, it is difficult to tell why `XX` or `YY` are set to `None`.

Reply:
I think that we would need a small comment to clarify this part.